### PR TITLE
Support n-dimensional empty tensors in CUDA non-reduction dimension f…

### PR DESCRIPTION
**Pull request by gchanan**
…unctions.

This also unifies the error checkign between scatter/scatterAdd on CUDA.

### CODE REVIEW

**Comment by gchanan** [comment]
I wasn't consistent about putting things in if blocks vs returning early; I generally prefer if blocks (even if causes history churn) because it's less likely to have bugs going forward.

File: aten/src/THC/THCTensorSort.cu [hunk 1]
```diff
@@ -3,60 +3,61 @@
...
-  int64_t dims = THCudaLongTensor__nDimension(state, t);
+  int64_t dims = THCudaLongTensor_nDimension(state, t);
   THArgCheck(dims <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
 
   ptrdiff_t inElements = THCudaLongTensor_nElement(state, t);
-  int64_t sliceSize = THCudaLongTensor_size(state, t, dim);
-  ptrdiff_t numSlices = inElements / sliceSize;
+  if (inElements > 0) {
+    int64_t sliceSize = THCudaLongTensor_size(state, t, dim);
+    ptrdiff_t numSlices = inElements == 0 ? 0 : inElements / sliceSize;
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/THC/THCTensorSort.cu]: yah, I moved this code around a few times, will fix.

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 2]
```diff
@@ -197,25 +201,30 @@ void THCTensor_(scatterAdd)(THCState* state, THCTensor *tensor, int dim, THCudaL
...
-  THLongStorage_free(indexDims);
 
-  for (int d = 0; d < THCTensor_(_nDimension)(state, tensor); d++) {
+  for (int d = 0; d < THCTensor_(nDimension)(state, tensor); d++) {
     if (d != dim) {
       THArgCheck(THCTensor_(size)(state, tensor, d) == THCTensor_(size)(state, src, d), 4,
                  "Input tensor must have same size as output tensor apart from the specified dimension");
     }
+    int64_t indexSizeD = THCudaLongTensor_size(state, index, d);
+    if (d != dim) {
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorScatterGather.cu]: You seem to be testing this case twice now.

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 3]
```diff
@@ -197,25 +201,30 @@ void THCTensor_(scatterAdd)(THCState* state, THCTensor *tensor, int dim, THCudaL
...
+      THArgCheck(indexSizeD <= THCTensor_(size)(state, tensor, d), 3,
+                 "Index tensor must not have larger size than output tensor apart from the specified dimension %d, but got index %s output %s",
+                 dim, THCudaLongTensor_sizeDesc(state, index).str, THCTensor_(sizeDesc)(state, tensor).str);
+    }
+    THArgCheck(indexSizeD <= THCTensor_(size)(state, src, d), 3,
+               "Index tensor must not have larger size than input tensor, but got index %s input %s",
+               THCudaLongTensor_sizeDesc(state, index).str, THCTensor_(sizeDesc)(state, src).str);
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorScatterGather.cu]: Which test exercises this error?

File: aten/src/THC/generic/THCTensorSort.cu [hunk 4]
```diff
@@ -13,20 +13,21 @@ THC_API void THCTensor_(sortKeyValueInplace)(THCState* state,
...
+  dims = THCTensor_(nDimension)(state, key);
   THArgCheck(dims <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
 
   ptrdiff_t inElements = THCTensor_(nElement)(state, key);
-  int64_t keySliceSize = THCTensor_(size)(state, key, dim);
-  ptrdiff_t keySlices = inElements / keySliceSize;
 
-  if (THCTensor_(_nDimension)(state, key) == 0) {
+  if (inElements == 0) {
     // Zero-dim tensor; do nothing
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorSort.cu]: Comment is out of date

**Review by ezyang — APPROVED** [review]
Looks good after nits.

**Comment by gchanan** [comment]
nits will be fixed in an upcoming PR.

**Pull request closed without merging**